databuilder/rest_api/rest_api_query.py

```diff
@@ -0,0 +1,203 @@
+
+    def __init__(self,
+                 query_to_join,  # type: BaseRestApiQuery
+                 url,  # type: str
+                 params,  # type: Dict[str, Any]
+                 json_path,  # type: str
+                 field_names,  # type: List[str]
+                 fail_no_result=False,  # type: bool
+                 skip_no_result=False,  # type: bool
+                 ):
+        # type: (...) -> None
+        """
+
+        :param query_to_join: Previous query to JOIN. RestApiQuerySeed can be used for the first query
+        :param url: URL string. It will use <str>.format operation using record that comes from previous query to
+        substitute any variable that URL has.
+        e.g: Previous record: {"dashboard_id": "foo"},
+             URL before: http://foo.bar/dashboard/{dashboard_id}
+             URL after compute: http://foo.bar/dashboard/foo
+
+        :param params: A keyword arguments that pass into requests.get function.
+            https://requests.readthedocs.io/en/master/user/quickstart/#make-a-request
+        :param json_path: A JSONPath expression. https://github.com/kennknowles/python-jsonpath-rw
+        Example:
+            JSON result:
+            [{"report_id": "1", "report_name": "first report", "foo": "bar"},
+             {"report_id": "2", "report_name": "second report"}]
+
+            JSON PATH:
+            [*].[report_id,report_name]
+
+            ["1", "first report", "2", "second report"]
+        :param field_names: Field names to be used on the result. Result is dictionary where field_name will be the key
+         and the values extracted via JSON PATH will be the value.
+
+            JSON Path result:
+                ["1", "first report", "2", "second report"]
+
+            field_names:
+                ["dashboard_id", "dashboard_description"]
+
+            {"dashboard_id": "1", "dashboard_description": "first report"}
+            {"dashboard_id": "2", "dashboard_description": "second report"}
+
+        :param fail_no_result: If there's no result from the query it will make it fail.
+        :param skip_no_result: If there's no result from the query, it will skip this record.
+        """
+        self._inner_rest_api_query = query_to_join
+        self._url = url
+        self._params = params
+        self._json_path = json_path
+        self._jsonpath_expr = parse(self._json_path)
+        self._fail_no_result = fail_no_result
+        self._skip_no_result = skip_no_result
+        self._field_names = field_names
+        self._more_pages = False
+
+    def execute(self):
+        # type: () -> Iterator[Dict[str, Any]]
+
+        self._authenticate()
+
+        for record_dict in self._inner_rest_api_query.execute():
+
+            first_try = True  # To control pagination. Always pass the while loop on the first try
+            while first_try or self._more_pages:
+                first_try = False
+
+                url = self._preprocess_url(record=record_dict)
```

Comment:
if I understand correctly, the max api join is to two apis? and it always expect the current url params is from the seed_query/query_to_join?

Reply:
Number of join can be numerous -- more than two.

On following example, it shows join of `5` RestQueryApi queries. In other words, it joins two `RestQueryApi` at a time, but we can keep chaining/decorating them.

```
RestQueryApi_A JOIN RestQueryApi_B
RestQueryApi_B JOIN RestQueryApi_C
RestQueryApi_C JOIN RestQueryApi_D
RestQueryApi_B JOIN RestQueryApi_E
...
```

Seed query is just to have `Decorator` pattern to get started because `RestQueryApi` always look for something to join. Hope this makes sense.